### PR TITLE
Tidy `modcc` driver, remove optimize flag.

### PR DESCRIPTION
**Pull request by halfflat**
* Remove optimization option (use SIMD options for vectorization).
* Remove arbor utility library dependencies from modcc (pending separation of utility lib from arbor lib source).
* Split target (cpu, gpu) specification from vectorization architecture (avx2, avx512).
* Remove `Options` singleton; replace with structure local to `modcc.cpp`.
* Tidy `modcc` option parsing and main function; allow a single invocation of `modcc` to generate code for multiple backends.
* Rename generated sources to include backend target in filename.
* Always run a constant simplification pass on generated procedures.
* Remove file i/o code from `Module` and `modcc` main function; move functionality to new functions in `io` namespace. (Note: in on-going mechanism revamp, other i/o utility code will reside in the `io` namespace and subdirectory.)
* Remove classes `ConstantFolderVisitor` and `ExpressionClassifierVisitor` that are no longer used.
* Modify CMakeLists.txt files, `backends/*/fvm.cpp` to reflect the new filenames of generated sources.
* Small formatting changes in `modcc` source to reflect coding guidelines (incomplete).

### CODE REVIEW

File: modcc/backends/base.hpp [hunk 1]
```diff
@@ -8,15 +8,18 @@
...
+enum class simdKind {
+    none, avx2, avx512
+};
+
 namespace modcc {
 
-using arb::util::enable_if_t;
+template <bool V, typename R = void>
+using enable_if_t = typename std::enable_if<V, R>::type;
+
```
Comment (vkarak) [inline modcc/backends/base.hpp]: I suppose that's related to the separation of the utility lib from the arbor source, right?

Comment (halfflat) [reply modcc/backends/base.hpp]: Exactly, yes. I plan to put it back once we have a separate utility source directory :)

**Comment by halfflat** [comment]
Argh. No `std::hash` specialization for enum defect strikes again. Will amend.

File: modcc/io/bulkio.hpp [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+// Read or write the contents of a file in toto.
+
+#include <string>
+#include <iterator>
+#include <fstream>
+
+namespace io {
+
+template <typename Assignable>
```
Comment (vkarak) [inline modcc/io/bulkio.hpp]: It looks like `Assignable` is a container. I see that your only requirement here is that `A` provides an `assign()` function, so perhaps you don't need another concept name, e.g., `SequenceContainer` etc. Would it make sense to have sth like `AssignableContainer`, 'cause initially I got a bit confused. I thought you need sth to support assignment and there are already a `CopyAssignable` and `MoveAssignable` concepts with a different meaning.

Comment (halfflat) [reply modcc/io/bulkio.hpp]: Certainly happy to rename! Perhaps `HasAssign`?

Comment (vkarak) [reply modcc/io/bulkio.hpp]: `HasAssign` sounds fine for me.

File: modcc/io/bulkio.hpp [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+// Read or write the contents of a file in toto.
+
+#include <string>
+#include <iterator>
+#include <fstream>
+
+namespace io {
+
+template <typename HasAssign>
+void snarf(std::istream& in, HasAssign& A) {
```
Comment (bcumming) [inline modcc/io/bulkio.hpp]: I appreciate that `snarf` and friends have history, but I would prefer to use something more descriptive for people who have been programming less than 20 years.

**Pull request merged**